ngraph/core/reference/include/ngraph/runtime/reference/roi_align.hpp

```diff
@@ -0,0 +1,239 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#pragma once
+
+#include <algorithm>
+#include "ngraph/coordinate_transform.hpp"
+#include "ngraph/op/roi_align.hpp" // for ROIAlign:PoolingMode
+#include "ngraph/shape.hpp"
+namespace ngraph
+{
+    namespace runtime
+    {
+        namespace reference
+        {
+            using ROIPoolingMode = op::v3::ROIAlign::PoolingMode;
+            template <typename T>
+            void roi_align(const T* feature_maps,
+                           const T* rois,
+                           const int64_t* batch_indices,
+                           T* out,
+                           const Shape& feature_maps_shape,
+                           const Shape& rois_shape,
+                           const Shape& batch_indices_shape,
+                           const Shape& out_shape,
+                           const int pooled_height,
+                           const int pooled_width,
+                           const int sampling_ratio,
+                           const float spatial_scale,
+                           const ROIPoolingMode& pooling_mode)
+            {
+                auto C = feature_maps_shape[1];
+                auto feature_map_height = feature_maps_shape[2];
+                auto feature_map_width = feature_maps_shape[3];
+                auto num_rois = rois_shape[0];
+
+                CoordinateTransform feature_maps_transform(feature_maps_shape);
+                CoordinateTransform rois_transform(rois_shape);
+                CoordinateTransform out_transform(out_shape);
+
+                for (unsigned int roi_index = 0; roi_index < num_rois; roi_index++)
+                {
+                    // Get ROI`s corners
+                    T x1 = rois[rois_transform.index({roi_index, 0})] * spatial_scale;
+                    T y1 = rois[rois_transform.index({roi_index, 1})] * spatial_scale;
+                    T x2 = rois[rois_transform.index({roi_index, 2})] * spatial_scale;
+                    T y2 = rois[rois_transform.index({roi_index, 3})] * spatial_scale;
+
+                    T roi_width = std::max(x2 - x1, static_cast<T>(1.0));
+                    T roi_height = std::max(y2 - y1, static_cast<T>(1.0));
+
+                    T bin_width = roi_width / pooled_width;
+                    T bin_height = roi_height / pooled_height;
+
+                    uint64_t num_samples_in_bin = sampling_ratio * sampling_ratio;
```

Comment:
In spec there is described case when `sampling_ratio == 0`. In such case we should use formula `ceil(roi_height / pooled_h)` and `ceil(roi_width / pooled_w)`

Reply:
You`re right